File: frontend/src/api/wizard.ts
import type { ApiResponse } from "@/types";
import { apiPost } from "./client";

// ─── Request Types ───────────────────────────────────────────────────

export interface WizardDescribeRequest {
  description: string;
  workspace_id?: string;
}

// ─── Response Types ──────────────────────────────────────────────────

export interface TemplateMatchResult {
  template_id: string;
  template_name: string;
  similarity: number;
}

export interface NLAnalysis {
  analysis_id: string;
  original_description: string;
  intents: string[];
  entities: string[];
  connectors_detected: string[];
  template_matches: TemplateMatchResult[];
  confidence: number;
}

export interface PlannedNode {
  node_id: string;
  label: string;
  node_type: string;
  description: string;
  config: Record<string, unknown>;
}

export interface PlannedEdge {
  source: string;
  target: string;
  condition: string;
}

export interface ModelRequirement {
  provider: string;
  model_id: string;
  purpose: string;
}

export interface ConnectorRequirement {
  connector_type: string;
  name: string;
  purpose: string;
}

export interface CredentialRequirement {
  vault_path: string;
  connector_type: string;
  description: string;
}

export interface NLBuildPlan {
  plan_id: string;
  analysis_id: string;
  nodes: PlannedNode[];
  edges: PlannedEdge[];
  models_needed: ModelRequirement[];
  connectors_needed: ConnectorRequirement[];
  credential_requirements: CredentialRequirement[];
  estimated_cost: number;
  security_notes: string[];
}

export interface GeneratedAgent {
  agent_name: string;
  tenant_id: string;
  workspace_id: string;
  owner_id: string;
  graph_definition: Record<string, unknown>;
  python_source: string;
  credential_manifest: CredentialRequirement[];
  plan_id: string;
}

export interface SecurityIssue {
  severity: string;
  code: string;
  message: string;
}

export interface ValidationResult {
  passed: boolean;
  security_issues: SecurityIssue[];
  cost_estimate: number;
  compliance_notes: string[];
}

export interface FullPipelineResult {
  agent: GeneratedAgent;
  validation: ValidationResult;
}

// ─── API Calls ───────────────────────────────────────────────────────
// TODO: Backend wizard.py has prefix="/api/v1/wizard" AND is registered with API_PREFIX — double prefix bug.
// Frontend paths below are correct in intent (/wizard/...) but won't resolve until backend is fixed.

/** Step 1: Describe — NLP analysis of a description */
export async function wizardDescribe(
  description: string,
  workspaceId?: string,
): Promise<ApiResponse<NLAnalysis>> {
  return apiPost<NLAnalysis>("/wizard/describe", {
    description,
    workspace_id: workspaceId,
  });
}

/** Step 2: Plan — structured build plan from analysis */
export async function wizardPlan(
  analysis: NLAnalysis,
): Promise<ApiResponse<NLBuildPlan>> {
  return apiPost<NLBuildPlan>("/wizard/plan", analysis);
}

/** Step 3: Build — generate agent from plan */
export async function wizardBuild(
  plan: NLBuildPlan,
): Promise<ApiResponse<GeneratedAgent>> {
  return apiPost<GeneratedAgent>("/wizard/build", plan);
}

/** Step 4: Validate — security scan & compliance check */
export async function wizardValidate(
  agent: GeneratedAgent,
): Promise<ApiResponse<ValidationResult>> {
  return apiPost<ValidationResult>("/wizard/validate", agent);
}

/** Refine — iterative refinement */
export async function wizardRefine(
  agent: GeneratedAgent,
  feedback: string,
  iteration: number,
): Promise<ApiResponse<GeneratedAgent>> {
  return apiPost<GeneratedAgent>("/wizard/refine", {
    agent,
    feedback,
    iteration,
  });
}

/** Full pipeline — describe → plan → build → validate */
export async function wizardFull(
  description: string,
  workspaceId?: string,
): Promise<ApiResponse<FullPipelineResult>> {
  return apiPost<FullPipelineResult>("/wizard/full", {
    description,
    workspace_id: workspaceId,
  });
}

File: backend/app/models/wizard.py
"""Pydantic models for the Natural Language → Agent Wizard pipeline."""

from __future__ import annotations

from typing import Any
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field


# ── Step 1: Describe ────────────────────────────────────────────────


class NLBuildRequest(BaseModel):
    """Incoming natural-language description of a desired agent."""

    description: str = Field(
        ...,
        min_length=1,
        max_length=5000,
        description="Free-text description of the agent to build",
    )
    workspace_id: str | None = Field(
        default=None,
        description="Optional workspace scope",
    )


class TemplateMatch(BaseModel):
    """A template that partially matches the user intent."""

    template_id: str
    template_name: str
    similarity: float = Field(ge=0.0, le=1.0)


class NLAnalysis(BaseModel):
    """Result of the *describe* step — NLP analysis of the request."""

    model_config = ConfigDict(populate_by_name=True)

    analysis_id: str = Field(default_factory=lambda: uuid4().hex)
    original_description: str = Field(alias="description")
    intents: list[str] = Field(default_factory=list)
    entities: list[str] = Field(default_factory=list)
    connectors_detected: list[str] = Field(default_factory=list)
    template_matches: list[TemplateMatch] = Field(default_factory=list)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)


# ── Step 2: Plan ────────────────────────────────────────────────────


class PlannedNode(BaseModel):
    """A single node in the planned agent graph."""

    node_id: str
    label: str
    node_type: str = "default"
    description: str = ""
    config: dict[str, Any] = Field(default_factory=dict)


class PlannedEdge(BaseModel):
    """A directed edge between two planned nodes."""

    source: str
    target: str
    condition: str = ""


class ModelRequirement(BaseModel):
    """An LLM model required by the planned agent."""

    provider: str
    model_id: str
    purpose: str = ""


class ConnectorRequirement(BaseModel):
    """An external connector required by the planned agent."""

    connector_type: str
    name: str
    purpose: str = ""


class CredentialRequirement(BaseModel):
    """A Vault credential path needed at runtime."""

    vault_path: str
    connector_type: str
    description: str = ""


class NLBuildPlan(BaseModel):
    """Result of the *plan* step — structured build plan."""

    plan_id: str = Field(default_factory=lambda: uuid4().hex)
    analysis_id: str
    nodes: list[PlannedNode] = Field(default_factory=list)
    edges: list[PlannedEdge] = Field(default_factory=list)
    models_needed: list[ModelRequirement] = Field(default_factory=list)
    connectors_needed: list[ConnectorRequirement] = Field(default_factory=list)
    credential_requirements: list[CredentialRequirement] = Field(default_factory=list)
    estimated_cost: float = 0.0
    security_notes: list[str] = Field(default_factory=list)


# ── Step 3: Build ───────────────────────────────────────────────────


class GeneratedAgent(BaseModel):
    """Result of the *build* step — a fully generated agent definition."""

    agent_name: str
    tenant_id: str
    workspace_id: str = ""
    owner_id: str = ""
    graph_definition: dict[str, Any] = Field(default_factory=dict)
    python_source: str = ""
    credential_manifest: list[CredentialRequirement] = Field(default_factory=list)
    plan_id: str = ""


# ── Step 4: Validate ────────────────────────────────────────────────


class SecurityIssue(BaseModel):
    """A single security finding from the validation scan."""

    severity: str = "warning"
    code: str = ""
    message: str = ""


class ValidationResult(BaseModel):
    """Result of the *validate* step."""

    passed: bool = True
    security_issues: list[SecurityIssue] = Field(default_factory=list)
    cost_estimate: float = 0.0
    compliance_notes: list[str] = Field(default_factory=list)


# ── Refine ──────────────────────────────────────────────────────────


class RefineRequest(BaseModel):
    """Request body for the iterative refinement endpoint."""

    agent: GeneratedAgent
    feedback: str = Field(..., min_length=1, max_length=2000)
    iteration: int = Field(default=1, ge=1, le=3)


__all__ = [
    "ConnectorRequirement",
    "CredentialRequirement",
    "GeneratedAgent",
    "ModelRequirement",
    "NLAnalysis",
    "NLBuildPlan",
    "NLBuildRequest",
    "PlannedEdge",
    "PlannedNode",
    "RefineRequest",
    "SecurityIssue",
    "TemplateMatch",
    "ValidationResult",
]
